### DISCUSSION TITLE
Home, End, PgUp, PgDown, Arrow keys not working in neovim inside kitty

### DISCUSSION THREAD

**Discussion by bauripalash** [kovidgoyal/kitty#5338] [Q&A] [answered]
I am quite new to kitty terminal. I have a minimal kitty and neovim setup.

Kitty.conf → https://github.com/bauripalash/mydotfiles/blob/main/.config/kitty/kitty.conf
Neovim config → https://github.com/bauripalash/mydotfiles/blob/main/.config/nvim/init.vim

Home button, end button, page up, down button and arrow keys are not working in neovim inside kitty.

I tried old `vi`, `vim`, `nano` inside kitty, all buttons are working fine, but not working in neovim.
I tried neovim with `konsole`, `alacritty`, these problems are not present there.
Outside neovim, all buttons work just fine.
If I try to use those keys in normal mode, nothing happens, I mean they don't do `anything` at all and in insert mode, on those key presses a space character gets inserted.

I tried Ctrl+V, in neovim to check which key codes are being sent, apparently those keys are sending **Space**

Help Please!

**Comment by page-down** [comment, marked as answer]
Start troubleshooting the problem without the configuration.

For example
```shell
kitty --config=NONE nvim -u NONE
```

These keys all work fine in the latest version of kitty and neovim.

Neovim has enabled support for the kitty keyboard protocol by default after version 0.7, however it is not fully supported.
In the latest version 0.7.2, some keys (e.g. the numpad Enter) were not working properly, this issue has been (partially) fixed in the master branch.

EDIT:
As you have already tried, compiling from source or using the nightly version can help solve the problem. Although some key combinations are still not supported.

**Reply by bauripalash** [reply]
I tried it, as you mentioned. But not working.

Kitty version : `kitty 0.25.2 created by Kovid Goyal`

Neovim version : 
```
NVIM v0.7.2
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by builduser

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Reply by page-down** [reply]
- Make sure `TERM` has not been modified anywhere and should be the default `xterm-kitty`.
- Make sure the terminfo database is in the correct location, either `xterm-kitty` installed to `~/.terminfo/x/` (also `~/.terminfo/78/` on macOS), or `/usr/share/terminfo/x/`, or as specified by `TERMINFO` env var.

https://sw.kovidgoyal.net/kitty/faq/

If you find that the terminfo database file specified by `TERMINFO` does not exist, then it is likely that some linux package has intentionally deleted the file. In this case, you need packages like `kitty-terminfo`,or try the official binary release, which can be installed via the official installer.

The problem is related to the terminal keyboard mode.
EDIT: The CSI u escape sequence is used by default in nvim 0.7+, but is buggy. The following explanation does not apply to kitty + nvim 0.7+.

For example, neovim is configured with key bindings for application mode, but switches to normal mode for any reason, so kitty sends the escape sequence for normal mode.

```shell
nvim -V3
```

```log
--- Terminal info --- {{{
&term: xterm-kitty
Description: KovIdTTY
  ...
  key_home                  khome      = <Esc>OH
  ...
  keypad_local              rmkx       = <Esc>[?1l
  keypad_xmit               smkx       = <Esc>[?1h
```

You can search and find out more details in neovim GitHub Issues.

**Reply by bauripalash** [reply]
TERM is set to xterm-kitty.
TERMINFO is also pointing to xterm-kitty in `/usr/share/terminfo/xterm-kitty`
There are some files in that directory like kitty.terminfo and some others. 

For the sake of experiment, I installed neovim nightly from AUR, and the problem is no longer present.

**Comment by alkaris2** [comment]
I'm having this very issues where the Page Up and Page Down keys wont work. Running `kitty 0.26.5`

So even though in the config file for Kitty I have

```
map page_up            scroll_page_up
map page_down          scroll_page_down
```

this does nothing, the keys will not work, however if I change it to the following;

```
map pgup            scroll_page_up
map pgdown          scroll_page_down
```

the key <kbd>PgDown</kbd> works, but <kbd>PgUp</kbd> doesn't. I tried changing the wording around by making it `pageup` or `page_up` that didn't work either, but if I changed it to `pg_up` then does it work as it should. It would seem the default wording `page_up` isn't what it should be then if it doesn't work like that. Is that is mistake of whoever made the default config or did something get changed and not set with the correct keycode mapping? It also could be quite possible the key mapping of certain key codes are different depending on the kind of keyboard you are using where one keyboard will read input from `page_up` as `pg_up` or `pgup`

**Reply by page-down** [reply]
> ... map page_up  ... this does nothing
> ... but if I changed it to pg_up then does it work as it should.

I assume that you want Neovim to receive the page up keypress and respond to it.
However, after configuring the correct key mapping, kitty will perform the action (scrollback up) and will not send that key to Neovim any longer.

So when you configure it with the wrong name, the keys will be sent to Neovim and processed.

For the `Page Up` key, the correct name is `page_up`, with some aliases. These are defined in the code.
https://github.com/kovidgoyal/kitty/blob/v0.26.5/kitty/key_names.py
https://sw.kovidgoyal.net/kitty/keyboard-protocol/#functional
